### PR TITLE
feat: Added check box to combine items with same BOM

### PR DESCRIPTION
**Pull request by noahjacob**
**Improvements**
- Added a check box to let user combine items with same BOM, which makes item-wise work order creation possible.
- Created a reference table to keep track of the combined item's sales order after combined.
- On submitting work order, work order qty in sales orders are updated using the reference table.
prod_plan
- **[Docs](https://github.com/frappe/erpnext_documentation/pull/336)**

### CODE REVIEW

File: erpnext/manufacturing/doctype/production_plan/production_plan.js [hunk 1]
```diff
@@ -2,21 +2,21 @@
 // For license information, please see license.txt
 
 frappe.ui.form.on('Production Plan', {
-	setup: function(frm) {
+	setup: function (frm) {
```
Comment (ankush) [inline erpnext/manufacturing/doctype/production_plan/production_plan.js]: Please keep cosmetic/formatting changes in separate commit.

File: erpnext/manufacturing/doctype/work_order/work_order.py [hunk 2]
```diff
@@ -241,7 +241,13 @@ def on_submit(self):
 		if not self.fg_warehouse:
 			frappe.throw(_("For Warehouse is required before Submit"))
 
-		self.update_work_order_qty_in_so()
+		prod_plan = frappe.get_doc('Production Plan', self.production_plan)
+		pp_ref = prod_plan.prod_plan_ref
+		if pp_ref:
```
Comment (rohitwaghchaure) [inline erpnext/manufacturing/doctype/work_order/work_order.py]: why extra variable just use ```if prod_plan.prod_plan_ref```

File: erpnext/manufacturing/doctype/work_order/work_order.py [hunk 3]
```diff
@@ -252,7 +258,13 @@ def on_cancel(self):
...
 
 		frappe.db.set(self,'status', 'Cancelled')
-		self.update_work_order_qty_in_so()
+		prod_plan = frappe.get_doc('Production Plan', self.production_plan)
+		pp_ref = prod_plan.prod_plan_ref
+		if pp_ref:
+			self.update_work_order_qty_in_combined_so(cancel = True)
+		else:
+			self.update_work_order_qty_in_so()
+
```
Comment (rohitwaghchaure) [inline erpnext/manufacturing/doctype/work_order/work_order.py]: Same code has added in the on_submit method better create separate method for this portion and call it from on_submit as well on_cancel.

File: erpnext/manufacturing/doctype/production_plan/production_plan.json [hunk 4]
```diff
@@ -316,13 +319,30 @@
    "fieldname": "include_safety_stock",
    "fieldtype": "Check",
    "label": "Include Safety Stock in Required Qty Calculation"
+  },
+  {
+   "fieldname": "prod_plan_ref",
```
Comment (rohitwaghchaure) [inline erpnext/manufacturing/doctype/production_plan/production_plan.json]: prod_plan_references?

File: erpnext/manufacturing/doctype/work_order/work_order.py [hunk 5]
```diff
@@ -357,6 +369,35 @@ def update_work_order_qty_in_so(self):
...
+			if not total_bundle_qty:
+				# product bundle is 0 (product bundle allows 0 qty for items)
+				total_bundle_qty = 1
+
+		prod_plan = frappe.get_doc('Production Plan', self.production_plan)
+		pp_item = frappe.get_doc('Production Plan Item', self.production_plan_item)
+		
+		for p in prod_plan.prod_plan_ref:
+			if p.item_ref == pp_item.item_reference:
```
Comment (rohitwaghchaure) [inline erpnext/manufacturing/doctype/work_order/work_order.py]:
```
work_order_qty = 0.0
if self.docstatus == 1:
  work_order_qty = cint(p.qty) / total_bundle_qty
```

File: erpnext/manufacturing/doctype/work_order/work_order.py [hunk 6]
```diff
@@ -241,7 +241,13 @@ def on_submit(self):
 		if not self.fg_warehouse:
 			frappe.throw(_("For Warehouse is required before Submit"))
 
-		self.update_work_order_qty_in_so()
+		prod_plan = frappe.get_doc('Production Plan', self.production_plan)
+		
+		if prod_plan.prod_plan_ref:
```
Comment (rohitwaghchaure) [inline erpnext/manufacturing/doctype/work_order/work_order.py]: if self.production_plan and frappe.db.exists('Production Plan Item Reference', {'parent': self.production_plan})

File: erpnext/manufacturing/doctype/work_order/work_order.py [hunk 7]
```diff
@@ -357,6 +369,35 @@ def update_work_order_qty_in_so(self):
...
+		if self.product_bundle_item:
+			total_bundle_qty = frappe.db.sql(""" select sum(qty) from
+				`tabProduct Bundle Item` where parent = %s""", (frappe.db.escape(self.product_bundle_item)))[0][0]
+
+			if not total_bundle_qty:
+				# product bundle is 0 (product bundle allows 0 qty for items)
+				total_bundle_qty = 1
+
+		prod_plan = frappe.get_doc('Production Plan', self.production_plan)
+		pp_item = frappe.get_doc('Production Plan Item', self.production_plan_item)
```
Comment (rohitwaghchaure) [inline erpnext/manufacturing/doctype/work_order/work_order.py]: Why get_doc, use prod_plan.items

File: erpnext/manufacturing/doctype/production_plan/production_plan.py [hunk 8]
```diff
@@ -165,10 +167,24 @@ def get_mr_items(self):
 		self.calculate_total_planned_qty()
 
 	def add_items(self, items):
-		self.set('po_items', [])
+		refs = {}
 		for data in items:
 			item_details = get_item_details(data.item_code)
+			if self.combine_items:	
+				if item_details.bom_no in refs.keys():
+					refs[item_details.bom_no]['qty'] = refs[item_details.bom_no]['qty'] + data.pending_qty
```
Comment (rohitwaghchaure) [inline erpnext/manufacturing/doctype/production_plan/production_plan.py]: refs[item_details.bom_no]['qty'] += data.pending_qty

**Comment by rohitwaghchaure** [comment]
@noahjacob Check sider issue

**Comment by noahjacob** [comment]
> @noahjacob Check sider issue

@rohitwaghchaure fixed

File: erpnext/manufacturing/doctype/production_plan/production_plan.py [hunk 9]
```diff
@@ -165,10 +167,24 @@ def get_mr_items(self):
...
+					refs[item_details.bom_no]['planned_qty'].append(data.pending_qty)
+					continue
+				else:
+					refs[item_details.bom_no] = {'qty': data.pending_qty, 'ref': data.name}
+					refs[item_details.bom_no]['so'] = [data.parent]
+					refs[item_details.bom_no]['so_items'] = [data.name]
+					refs[item_details.bom_no]['planned_qty'] = [data.pending_qty]
+		
 			pi = self.append('po_items', {
+				'name': data.name,
```
Comment (rohitwaghchaure) [inline erpnext/manufacturing/doctype/production_plan/production_plan.py]: why this line added?

File: erpnext/manufacturing/doctype/production_plan/production_plan.py [hunk 10]
```diff
@@ -165,10 +167,24 @@ def get_mr_items(self):
...
+				if item_details.bom_no in refs.keys():
+					refs[item_details.bom_no]['qty'] += data.pending_qty
+					refs[item_details.bom_no]['so'].append(data.parent)	
+					refs[item_details.bom_no]['so_items'].append(data.name)
+					refs[item_details.bom_no]['planned_qty'].append(data.pending_qty)
+					continue
+				else:
+					refs[item_details.bom_no] = {'qty': data.pending_qty, 'ref': data.name}
+					refs[item_details.bom_no]['so'] = [data.parent]
+					refs[item_details.bom_no]['so_items'] = [data.name]
```
Comment (rohitwaghchaure) [inline erpnext/manufacturing/doctype/production_plan/production_plan.py]: data.name used in line no 181 as well on 183, better you create a structure with tuple like as below

```
key = ('sales_order', 'bom_no')

# data.name is the reference of po_items
refs = {key: ['data.name', 'data1.name']}
```

File: erpnext/manufacturing/doctype/work_order/work_order.py [hunk 11]
```diff
@@ -357,7 +365,28 @@ def update_work_order_qty_in_so(self):
...
+				total_bundle_qty = 1
 
+		prod_plan = frappe.get_doc('Production Plan', self.production_plan)
+		item_reference = frappe.get_value('Production Plan Item', self.production_plan_item, 'sales_order_item')
+		
+		for plan_reference in prod_plan.prod_plan_references:
+			work_order_qty = 0.0
+			if plan_reference.item_reference == item_reference:
+				if self.docstatus == 1:
+					work_order_qty = cint(plan_reference.qty) / total_bundle_qty
```
Comment (rohitwaghchaure) [inline erpnext/manufacturing/doctype/work_order/work_order.py]: why cint and not flt?

**Review by rohitwaghchaure — APPROVED** [review]

**Pull request merged**